### PR TITLE
Add GCC to PR gate

### PR DESCRIPTION
**Pull request by franksinankaya**
@am11 @jkotas @janvorli 

I don't know if this is going to work or not but I copied what clang does.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @dotnet/coreclr-infra

**Comment by AaronRobinsonMSFT** [comment]
@franksinankaya Is the intent here simply to have GCC support? Adding this to the pipeline is probably something we aren't really interested in doing since clang is what we officially support. Aside from that ensuring the CMake files have valid GCC options and that the code is acceptable to GCC seems reasonable if that is how you would like to limit the PR.

/cc @jkotas

**Comment by franksinankaya** [comment]
Yes, we want to add GCC support. Coreclr baseline already supports GCC now. This has been discussed here:

https://github.com/dotnet/coreclr/issues/24170

**Comment by AaronRobinsonMSFT** [comment]
@franksinankaya Ah I see the related threads. As mentioned in https://github.com/dotnet/coreclr/issues/23386#issuecomment-544817180 we are nearing completion of the consolidation effort, but aren't quite ready to take these types of PRs yet. The CI system is going to be in significant flux while addressing issues with the consolidation. Once the consolidated repo is up and stable these kind of PRs can be considered.

File: eng/pipelines/pr.yml [hunk 1]
```diff
@@ -95,6 +96,7 @@ jobs:
     - Linux_arm64
     - Linux_musl_x64
     - Linux_x64
+    - Linux_gcc_x64
```
Comment (jashook) [inline eng/pipelines/pr.yml]: Do we want to run tests for the gcc build?

File: eng/pipelines/pr.yml [hunk 2]
```diff
@@ -113,6 +115,7 @@ jobs:
     buildConfig: checked
     platforms:
     - Linux_x64
+    - Linux_gcc_x64
```
Comment (jashook) [inline eng/pipelines/pr.yml]: I definitely do not think we need to run r2r.

File: eng/pipelines/pr.yml [hunk 3]
```diff
@@ -131,6 +134,7 @@ jobs:
...
```
Comment (jashook) [inline eng/pipelines/pr.yml]: I do not think we should run corefx testing either. It is generally costly.

File: eng/platform-matrix.yml [hunk 4]
```diff
@@ -160,6 +160,38 @@ jobs:
...
+      archType: x64
+      osGroup: Linux
+      osSubgroup: _gcc
+      managedTestBuildOsGroup: OSX
+      container:
+        image: centos-7-f39df28-20191023143754
+        registry: mcr
+      helixQueues:
+      - ${{ if and(eq(variables['System.TeamProject'], 'public'), in(parameters.helixQueueGroup, 'pr', 'ci', 'corefx')) }}:
+        - Ubuntu.1804.Amd64.Open
```
Comment (jashook) [inline eng/platform-matrix.yml]: If we end up running testing for gcc build. I would suggest only running one platform. The full set is generally expensive.

File: eng/install-native-dependencies.sh [hunk 5]
```diff
@@ -5,7 +5,7 @@ if [ "$1" = "Linux" ]; then
     if [ "$?" != "0" ]; then
        exit 1;
     fi
-    sudo apt install cmake llvm-3.9 clang-3.9 lldb-3.9 liblldb-3.9-dev libunwind8 libunwind8-dev gettext libicu-dev liblttng-ust-dev libcurl4-openssl-dev libssl-dev libkrb5-dev libnuma-dev
+    sudo apt install cmake gcc g++ llvm-3.9 clang-3.9 lldb-3.9 liblldb-3.9-dev libunwind8 libunwind8-dev gettext libicu-dev liblttng-ust-dev libcurl4-openssl-dev libssl-dev libkrb5-dev libnuma-dev
```
Comment (jashook) [inline eng/install-native-dependencies.sh]: This is not necessary. We do all build now in docker. I suggest doing the same.

**Comment by franksinankaya** [comment]
OK, please let me know when it is ready.
I assumed that I could move this into the same bucket as _musl.
When you move the _musl to the new system, you could move _gcc together.

We can as well wait for CI/CD system to be finished too.

**Comment by AaronRobinsonMSFT** [comment]
@jashook Do we have list of existing CI issues that we can track post consolidation?

**Comment by franksinankaya** [comment]
Initial goal is to build and test with gcc on ubuntu 18.04 only.

File: eng/pipelines/pr.yml (hunk 1 above)
Comment (franksinankaya) [reply eng/pipelines/pr.yml]: yes

File: eng/install-native-dependencies.sh (hunk 5 above)
Comment (franksinankaya) [reply eng/install-native-dependencies.sh]: done

**Comment by jkotas** [comment]
> Initial goal is to build and test with gcc on ubuntu 18.04 only.

We do not need to run all tests for gcc in CI. It should be a small smoke test at most (e.g. print hello world, throw and catch exception, and trigger a GC). The chances that we see gcc specific bug is very low.

Note that we do not have a full matrix coverage in CI. To reduce costs and increase reliability, we run checked or release build only for some flavors, etc. We will be even more strategic about the combinations that CI runs in the unified repo.

**Comment by jkotas** [comment]
> The CI system is going to be in significant flux

Yep, feel free to experiment and gather feedback here, but do not expect this PR to be merged into coreclr repo.

**Comment by jashook** [comment]
>We do not need to run all tests for gcc in CI. It should be a small smoke test at most (e.g. print hello world, throw and catch exception, and trigger a GC). The chances that we see gcc specific bug is very low.

We currently do not have an exact mechanism to do this. Maybe it makes sense to redefine our pri0/pri1 tests to potentially allow smoke testing maybe with pri0 (smoke tests), pri1 innerloop, pri2?

**Comment by jkotas** [comment]
I think the smoke test can be done as one-off. It should be just a single .cs with less than 1000 lines that compiles into a single .exe.

**Comment by franksinankaya** [comment]
Is there any value in sending a PR to the dockerfile to include gcc? or do you expect that to change too?

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

**Comment by franksinankaya** [comment]
awesome, looking forward to it.

**Pull request closed without merging**